src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/Cas1SeedPremisesFromSiteSurveyXlsxJob.kt

```diff
@@ -0,0 +1,289 @@
+    val probationRegion: ProbationRegionEntity,
+    val localAuthorityArea: LocalAuthorityAreaEntity,
+    val characteristicEntities: List<CharacteristicEntity>,
+    val longitude: Double,
+    val latitude: Double,
+    var point: Point,
+  )
+
+  private fun createPremise(
+    premisesInfo: PremisesInfo,
+  ) {
+    val qCode = premisesInfo.siteSurveyPremise.qCode
+    log.info("Creating new premise for qcode $qCode (${premisesInfo.siteSurveyPremise.name})")
+
+    val siteSurvey = premisesInfo.siteSurveyPremise
+
+    val approvedPremises = premisesRepository.save(
+      ApprovedPremisesEntity(
+        id = UUID.randomUUID(),
+        name = siteSurvey.name,
+        addressLine1 = siteSurvey.address,
+        addressLine2 = null,
+        town = siteSurvey.townCity,
+        postcode = siteSurvey.postcode,
+        notes = "",
+        // A new row is required in site surveys to capture this
+        emailAddress = null,
+        probationRegion = premisesInfo.probationRegion,
+        localAuthorityArea = premisesInfo.localAuthorityArea,
+        bookings = mutableListOf(),
+        lostBeds = mutableListOf(),
+        // A new row is required in site surveys to capture this
+        apCode = qCode,
+        qCode = qCode,
+        rooms = mutableListOf(),
+        characteristics = premisesInfo.characteristicEntities.toMutableList(),
+        status = PropertyStatus.active,
+        longitude = premisesInfo.longitude,
+        latitude = premisesInfo.latitude,
+        point = premisesInfo.point,
+        gender = siteSurvey.maleFemale.toApprovedPremisesGender(),
+        supportsSpaceBookings = false,
+        // A new row is required in site surveys to capture this
+        managerDetails = null,
+      ),
+    )
+
+    premisesRepository.save(approvedPremises)
+  }
+
+  private fun updatePremise(
+    existingPremise: ApprovedPremisesEntity,
+    premisesInfo: PremisesInfo,
+  ) {
+    val qCode = premisesInfo.siteSurveyPremise.qCode
+    log.info("Updating premise for qcode $qCode (${premisesInfo.siteSurveyPremise.name})")
+
+    val beforeChange = ApprovedPremisesForComparison.fromEntity(existingPremise)
+
+    val siteSurvey = premisesInfo.siteSurveyPremise
+    existingPremise.apply {
+      name = siteSurvey.name
+      addressLine1 = siteSurvey.address
+      addressLine2 = null
+      town = siteSurvey.townCity
+      postcode = siteSurvey.postcode
+      probationRegion = premisesInfo.probationRegion
+      localAuthorityArea = premisesInfo.localAuthorityArea
+      longitude = premisesInfo.longitude
+      latitude = premisesInfo.latitude
+      point = premisesInfo.point
+      gender = siteSurvey.maleFemale.toApprovedPremisesGender()
+    }
+
+    existingPremise.characteristics.clear()
+    existingPremise.characteristics.addAll(premisesInfo.characteristicEntities)
+
+    val afterChange = ApprovedPremisesForComparison.fromEntity(existingPremise)
+
+    val diff = javers.compare(beforeChange, afterChange)
```

Comment:
Nice - looks like a useful library

Reply:
Yes. If we refactored the room seed job to use an internal model before persisting (some data classes) we could then load the current entities into the same internal classes and generate a diff. This is less important in the short term as SWSC don't have any rooms defined to modify